arangod/IResearch/IResearchCalculationAnalyzer.cpp

```diff
@@ -0,0 +1,507 @@
+  //        }
+
+  //        result = createNodeAttributeAccess(result, part->getStringValue(),
+  //          part->getStringLength());
+  //      }
+
+  //      if (result != nullptr) {
+  //        return result;
+  //      }
+  //    }
+  //    // fallthrough to exception
+
+  //    // if no string value was inserted for the parameter name, this is an
+  //    // error
+  //    THROW_ARANGO_EXCEPTION_PARAMS(TRI_ERROR_QUERY_BIND_PARAMETER_TYPE,
+  //      node->getString().c_str());
+  //  } else if (node->type == NODE_TYPE_TRAVERSAL) {
+  //    extractCollectionsFromGraph(node->getMember(2));
+  //  } else if (node->type == NODE_TYPE_SHORTEST_PATH ||
+  //    node->type == NODE_TYPE_K_SHORTEST_PATHS) {
+  //    extractCollectionsFromGraph(node->getMember(3));
+  //  }
+
+  //  return node;
+  //};
+  //Ast::traverseAndModify(_ast->root(), func);
+}
+
+bool CalculationAnalyzer::next() {
+  if (_queryResults.get()) {
+    // TODO: call engine refill!
+    while (_queryResults->numRows() > _resultRowIdx) {
+      AqlValue const& value = _queryResults->getValueReference(_resultRowIdx++, _engine->resultRegister());
+      if (value.isString()) {
+        arangodb::velocypack::Builder builder;
+        value.toVelocyPack(&arangodb::velocypack::Options::Defaults, builder, true, false);
+        _str = builder.slice().copyString();
+        _term.value = irs::ref_cast<irs::byte_type>(irs::string_ref(_str));
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+bool CalculationAnalyzer::reset(irs::string_ref const& field) noexcept {
+  if (!_query) { // lazy initialization 
+    _query = std::make_unique<CalculationQueryContext>(*_calculationVocbase);
+    _ast = std::make_unique<Ast>(*_query);
+    // important to hold a copy here as parser accepts reference!
+    auto queryString = arangodb::aql::QueryString(_options.queryString);
+    Parser parser(*_query, *_ast, queryString);
+    parser.parse();
+    AstNode* astRoot = const_cast<AstNode*>(_ast->root());
+    TRI_ASSERT(astRoot);
+    Ast::traverseAndModify(astRoot, [this, field](AstNode * node) -> AstNode* {
+      if (node->type == NODE_TYPE_PARAMETER) {
+        // should be only our parameter name. see validation method!
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+        TRI_ASSERT(irs::string_ref(node->getStringValue(), 
+                                   node->getStringLength()) == CALCULATION_PARAMETER_NAME);
+#endif
+        // FIXME: move to computed value once here could be not only strings
+        auto newNode = _ast->createNodeValueString(field.c_str(), field.size());
+        _bindedNodes.push_back(newNode);
+        return newNode;
+      } else {
+        return node;
+      }
+      });
+    _plan = ExecutionPlan::instantiateFromAst(_ast.get());
+  } else {
+    for (auto node : _bindedNodes) {
+      node->setStringValue(field.c_str(), field.size());
+    }
+  }
+
+  _engine = std::make_unique<ExecutionEngine>(0, *_query, _query->itemBlockManager(),
```

Comment:
it seems that we can avoid using `unique_ptr` as `QueryCalculationContext` can be already initialized.